### PR TITLE
L1T uGT: Three body invariant mass for muons

### PR DESCRIPTION
**Pull request by elfontan**
#### PR description:

Development of a new condition class to deal with a three-body correlation condition (starting from CorrCondition and CorrWithOverlapRemovalCondition as a model): CorrThreeBodyCondition and CorrelationThreeBodyTemplate.
Based on the version of the UTM libraries utm_0.8.1: https://gitlab.cern.ch/cms-l1t-utm/utm/-/tree/utm_0.8.1.
New parseCorrelation function needed in the TriggerMenuParser. Minor changes needed in the L1TGlobalProducer, GlobalBoard, TruggerMenu, GlobalCondition, GlobalDefinitions.

#### PR validation:
    
1) Tests performed by producing test vectors using a dedicated L1 Menu with four algorithms requiring three muons with different pT cuts and a threshold on the three-muon invariant mass: successful comparison with triggers from the firmware.
CMSSW release used for tests: CMSSW_12_0_X_2021-05-24-2300.
2) Test at the HLT level to verify that the correct information about the objects that caused the decision is correctly saved and used: a generic triple muon HLT path can be seeded by L1 triggers with a three-body invariant mass requirement.
STEPS:
        - Private production of a small RAW-DIGI ZZTo4Mu sample (100 events) using Run 3 conditions 
        - Production of a dedicated HLT menu in ConfDB including FirstPath, FinalPath, and a generic TripleMu seed (HLT_TripleMu_10_5_5_DZ_v10) with modified L1 seeds as Logical Expression:
/users/elfontan/ThreeMuInvMassTest/TriMuInvarianMass1120/V4 
        - HLT configuration and uGT emulator run with a dedicated L1 test menu:
[... --l1-emulator uGT --mc --l1Xml L1Menu_test_mass_3_body_reduced_v2.xml"] 
3) Some final checks:
        - scram build code-format #Ensure that the code conforms to CMSSW standards 
        - scram b -j8 
        - runTheMatrix.py -l 11634.0

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc [hunk 1]
```diff
@@ -2810,6 +2823,151 @@ bool l1t::TriggerMenuParser::parseCorrelation(tmeventsetup::esCondition corrCond
...
+  if (objects.size() != 3) {
+    edm::LogError("TriggerMenuParser") << "incorrect number of objects for the correlation condition " << name
+                                       << " corrFlag " << corrFlag << std::endl;
+    return false;
+  }
+
+  // Loop over legs
+  for (size_t lll = 0; lll < objects.size(); lll++) {
+    const esObject object = objects.at(lll);
+    //std::cout << "      obj name = " << object.getName() << "\n";
```
[the remark this answers is not in the document]
Comment (elfontan) [reply L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc:2919]: Hi @qliphy , thank you for this, I missed them! I am going to commit changes to fix this now

**Comment by qliphy** [comment]
thanks @elfontan 
@cecilecaillol  Would you mind check and sign again if the changes looks ok to you?

**Comment by cecilecaillol** [comment]
+l1

**Comment by qliphy** [comment]
+1

**Pull request merged**